core-api/src/test/java/com/optimizely/ab/OptimizelyUserContextTest.java

```diff
@@ -1445,12 +1533,39 @@ public void findValidatedForcedDecisionWithoutRuleKey() {
             userId,
             Collections.emptyMap());
 
-        optimizelyUserContext.setForcedDecision(flagKey, variationKey);
-        DecisionResponse<Variation> response = optimizelyUserContext.findValidatedForcedDecision(flagKey);
+        OptimizelyDecisionContext optimizelyDecisionContext = new OptimizelyDecisionContext(flagKey, null);
+        OptimizelyForcedDecision optimizelyForcedDecision = new OptimizelyForcedDecision(variationKey);
+
+        optimizelyUserContext.setForcedDecision(optimizelyDecisionContext, optimizelyForcedDecision);
+        DecisionResponse<Variation> response = optimizelyUserContext.findValidatedForcedDecision(optimizelyDecisionContext);
         Variation variation = response.getResult();
         assertEquals(variationKey, variation.getKey());
     }
 
+    @Test
+    public void setForcedDecisionsAndCallDecide() {
+        String flagKey = "feature_2";
+        String ruleKey = "exp_no_audience";
+        String variationKey = "variation_with_traffic";
+        OptimizelyUserContext optimizelyUserContext = new OptimizelyUserContext(
+            optimizely,
+            userId,
+            Collections.emptyMap());
+
+        OptimizelyDecisionContext optimizelyDecisionContext = new OptimizelyDecisionContext(flagKey, ruleKey);
+        OptimizelyForcedDecision optimizelyForcedDecision = new OptimizelyForcedDecision(variationKey);
+        optimizelyUserContext.setForcedDecision(optimizelyDecisionContext, optimizelyForcedDecision);
+        assertEquals(variationKey, optimizelyUserContext.getForcedDecision(optimizelyDecisionContext).getVariationKey());
+
+        // Test to confirm decide uses proper FD
+        OptimizelyDecision decision = optimizelyUserContext.decide(flagKey, Arrays.asList(OptimizelyDecideOption.INCLUDE_REASONS));
+
+        assertNotNull(decision);
+        assertTrue(decision.getReasons().contains(
+            String.format("Variation (%s) is mapped to flag (%s), rule (%s) and user (%s) in the forced decision map.", variationKey, flagKey, ruleKey, userId)
+        ));
+    }
```

Comment:
Can we make this test separately for each of F->D, E->D, and D->D?
And for each case, validate impression event and notification fired with the correct contents.

Reply:
Added 3 new tests with each and creating the impression events and verifying the contents.